lib/deimos/active_record_consumer.rb

```diff
@@ -1,53 +1,64 @@
-require 'deimos/schema_model_converter'
 
 module Deimos
-  # Consumer that automatically saves the payload into the database.
+  # Basic ActiveRecord consumer class. Consumes messages and upserts them to
+  # the database. For tombstones (null payloads), deletes corresponding
+  # records from the database. Can operate in either message-by-message mode
+  # or in batch mode.
+  #
+  # In batch mode, ActiveRecord callbacks will be skipped and messages will
+  # be batched to minimize database calls.
+
+  # To configure batch vs. message mode, change the delivery mode of your
+  # Phobos listener.
+  # Message-by-message -> use `delivery: message` or `delivery: batch`
+  # Batch -> use `delivery: inline_batch`
   class ActiveRecordConsumer < Consumer
-    include ActiveRecordBaseConsumer
+    include ActiveRecord::MessageConsumption
+    include ActiveRecord::BatchConsumption
+
+    class << self
+      # param klass [Class < ActiveRecord::Base] the class used to save to the
+      # database.
+      def record_class(klass)
+        config[:record_class] = klass
+      end
+    end
 
-    # Create new consumer
+    # Setup
     def initialize
-      init_consumer
+      @klass = self.class.config[:record_class]
 
-      super
-    end
+      @converter = SchemaModelConverter.new(
+        self.class.decoder.avro_schema,
+        @klass
+      )
 
-    # :nodoc:
-    def consume(payload, metadata)
-      key = metadata.with_indifferent_access[:key]
-      record = @klass.unscoped.where(@klass.primary_key => key).first
-      if payload.nil?
-        destroy_record(record)
-        return
-      end
-      record ||= @klass.new
-      attrs = record_attributes(payload.with_indifferent_access)
-      # don't use attributes= - bypass Rails < 5 attr_protected
-      attrs.each do |k, v|
-        record.send("#{k}=", v)
+      if self.class.config[:key_schema] # rubocop:disable Style/GuardClause
+        @key_converter = SchemaModelConverter.new(
+          self.class.key_decoder.avro_schema,
+          @klass
+        )
       end
-      record[@klass.primary_key] = key
-      record.created_at ||= Time.zone.now if record.respond_to?(:created_at)
-      record.updated_at ||= Time.zone.now if record.respond_to?(:updated_at)
-      record.save!
     end
 
-    # Destroy a record that received a null payload. Override if you need
-    # to do something other than a straight destroy (e.g. mark as archived).
-    # @param record [ActiveRecord::Base]
-    def destroy_record(record)
-      record&.destroy
-    end
+  protected
+
+    # Get attributes for new/upserted records in the database. Override this
+    # method (with super) to customize the set of attributes used to instantiate
+    # records.
+    # @param payload [Hash] The decoded message payload.
+    # @param key [Hash] The decoded message key.
+    # @return [Hash|nil] Attribute set for the upserted record. nil to skip
+    # insertion.
+    def record_attributes(key, payload)
```

Comment:
Can we change this to `record_attributes(payload, key=nil)` to preserve backwards compatibility?